### DISCUSSION TITLE
مشکل کوئری محاسبه حجم مصرفی ادمین ها

### DISCUSSION THREAD

**Discussion by ItsOrv** [Gozargah/Marzban#1268] [Q&A]
توی این کوئری که برای محاسبه حجم total هر ادمین استفاده میشه فکر کنم مشکلی وجود داره:
```
SELECT admins.username, (SUM(users.used_traffic) + IFNULL(SUM(user_usage_logs.used_traffic_at_reset), 0)) / 1073741824
FROM admins 
LEFT JOIN users ON users.admin_id = admins.id 
LEFT JOIN user_usage_logs ON user_usage_logs.user_id = users.id
Group By admins.username
```
مشکلش اینه حجم total یوزر هایی که نامحدود هستن یا لیمیتد شدن رو اشتباه نشون میده و فکر کنم اون حجم بعد اخرین ریست رو یبار دیگه هم باهاش جمع میکنه یا اون مقدار حجم هم توی used_traffic_at_reset هست هم توی used_traffic پس دوبار جمع میشه
سعی کردم با تغییر کوعری یبار این حجم دومی رو از کاربرایی که status اونا limited هست حساب نکنم  ولی اینجوری میاد کمتر از حجم استفاده شده واقعی نشون میده

من شک کردم که داره اشتباه میزنه رفتم یه پنلی که کم مصرف شده رو دستی جمع زدم دیدم مشکل داره

بعد اومدم با این کوئری تک تک یوزر های یه ادمین رو خروجی گرفتم مقایسه کنم:
```
SELECT 
    users.username, 
    (SUM(users.used_traffic) + IFNULL(SUM(user_usage_logs.used_traffic_at_reset), 0)) / 1073741824 AS total_traffic_in_gb 
FROM 
    users 
LEFT JOIN 
    user_usage_logs ON user_usage_logs.user_id = users.id 
JOIN 
    admins ON users.admin_id = admins.id 
WHERE 
    admins.username = 'admin' 
GROUP BY 
    users.username;
```
خروجی به این صورت شد، مثلا با مقایسه voilet میشه این موضوع رو متوجه شد
![IMG_20240826_174942_854](https://github.com/user-attachments/assets/7e2ab174-e03a-41ff-99bc-cf7a770edbdd)

**Comment by alirezax5** [comment]
اگه میخواید حجم مصرف شده user_usage_logs نمایش نده از sql زیر استفاده کن.
```sql
SELECT 
    users.username, 
    (SUM(users.used_traffic)) / 1073741824 AS total_traffic_in_gb 
FROM 
    users 
JOIN 
    admins ON users.admin_id = admins.id 
WHERE 
    admins.username = 'admin' 
GROUP BY 
    users.username;
```

و نمیدونم user_usage_logs  رکورداش بعد از ریست شدن ثبت میشه یا کلا محاسبه میشه , اگه بعد از ریست شدن باشه کد مشکلی نداره اما اگه کلی از اول ساخت ثبت میشه کد دچار مشکله و باید فقط user_usage_logs   محاسبه بشه.

**Reply by ItsOrv** [reply]
الان خروجی به این صورت شد :
![image](https://github.com/user-attachments/assets/77e7c40a-d08d-46da-a52b-cd1fb1b67ee4)
یوزر voilet رو مقایسه کنید
ولی من میخوام دقیق اون مقدار total که توی پنل نمایش داده میشه نشون داده بشه و توی کوئری اصلی بتونم حجم مصرفی یک ادمین رو ببینم
نمیدونم بقیه که پنل میفروشن از چه کوئری استفاده میکنن که به مشکل نمیخورن، شایدم به مشکل میخورن و متوجه نشدن

**Reply by alirezax5** [reply]
> الان خروجی به این صورت شد : ![image](https://private-user-images.githubusercontent.com/70858122/361519150-77e7c40a-d08d-46da-a52b-cd1fb1b67ee4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ2OTM1NzcsIm5iZiI6MTcyNDY5MzI3NywicGF0aCI6Ii83MDg1ODEyMi8zNjE1MTkxNTAtNzdlN2M0MGEtZDA4ZC00NmRhLWE1MmItY2QxZmIxYjY3ZWU0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI2VDE3Mjc1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQzZDc3YWY3OTc0NjIyNWQwM2M0YjFjY2YzNTViYWVmNTkyNzBlM2IzZGM0Y2EyMmU0OGFiZWMxNGY4M2E3M2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.G_zYxTauNjQvBh2OEQL55RTJaIbaCBD3Y19b9hKFQ1E) یوزر voilet رو مقایسه کنید ولی من میخوام دقیق اون مقدار total که توی پنل نمایش داده میشه نشون داده بشه و توی کوئری اصلی بتونم حجم مصرفی یک ادمین رو ببینم نمیدونم بقیه که پنل میفروشن از چه کوئری استفاده میکنن که به مشکل نمیخورن، شایدم به مشکل میخورن و متوجه نشدن

میخوای فقط نام ادمین ، حجم مصرفی باشه؟

**Reply by ItsOrv** [reply]
این کوئری دومی رو فقط برای این نوشتم که دقیقتر چک کنم ببینم مشکل از کجاست، که دیدم حجم کاربرای لیمیت و اکسپایر شده رو اشتباه نشون میده
در اصل میخوام این کوئری که توی داکیومنت خود مرزبان هم هست مشکلش حل بشه و اشتباه نشون نده حجمارو:
```
SELECT admins.username, (SUM(users.used_traffic) + IFNULL(SUM(user_usage_logs.used_traffic_at_reset), 0)) / 1073741824
FROM admins 
LEFT JOIN users ON users.admin_id = admins.id 
LEFT JOIN user_usage_logs ON user_usage_logs.user_id = users.id
Group By admins.username
‍‍‍

**Comment by H-Alireza** [comment]
> توی این کوئری که برای محاسبه حجم total هر ادمین استفاده میشه فکر کنم مشکلی وجود داره:
> 
> ```
> SELECT admins.username, (SUM(users.used_traffic) + IFNULL(SUM(user_usage_logs.used_traffic_at_reset), 0)) / 1073741824
> FROM admins 
> LEFT JOIN users ON users.admin_id = admins.id 
> LEFT JOIN user_usage_logs ON user_usage_logs.user_id = users.id
> Group By admins.username
> ```
> 
> مشکلش اینه حجم total یوزر هایی که نامحدود هستن یا لیمیتد شدن رو اشتباه نشون میده و فکر کنم اون حجم بعد اخرین ریست رو یبار دیگه هم باهاش جمع میکنه یا اون مقدار حجم هم توی used_traffic_at_reset هست هم توی used_traffic پس دوبار جمع میشه سعی کردم با تغییر کوعری یبار این حجم دومی رو از کاربرایی که status اونا limited هست حساب نکنم ولی اینجوری میاد کمتر از حجم استفاده شده واقعی نشون میده
> 
> من شک کردم که داره اشتباه میزنه رفتم یه پنلی که کم مصرف شده رو دستی جمع زدم دیدم مشکل داره
> 
> بعد اومدم با این کوئری تک تک یوزر های یه ادمین رو خروجی گرفتم مقایسه کنم:
> 
> ```
> SELECT 
>     users.username, 
>     (SUM(users.used_traffic) + IFNULL(SUM(user_usage_logs.used_traffic_at_reset), 0)) / 1073741824 AS total_traffic_in_gb 
> FROM 
>     users 
> LEFT JOIN 
>     user_usage_logs ON user_usage_logs.user_id = users.id 
> JOIN 
>     admins ON users.admin_id = admins.id 
> WHERE 
>     admins.username = 'admin' 
> GROUP BY 
>     users.username;
> ```
> 
> خروجی به این صورت شد، مثلا با مقایسه voilet میشه این موضوع رو متوجه شد ![IMG_20240826_174942_854](https://private-user-images.githubusercontent.com/70858122/361474615-7e2ab174-e03a-41ff-99bc-cf7a770edbdd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU1NzgyMDQsIm5iZiI6MTcyNTU3NzkwNCwicGF0aCI6Ii83MDg1ODEyMi8zNjE0NzQ2MTUtN2UyYWIxNzQtZTAzYS00MWZmLTk5YmMtY2Y3YTc3MGVkYmRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTA1VDIzMTE0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNkOWYzODQwNWVjYzRkZDNkNjAxYWZjYTRhNTgxYjRhYjg3MzM0ZmI2NzFiYzE5Yzc3OWUyMGRmYjA2NzcwYzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.kgD9B6ouk6v_saiSEuW2pN9-4nzsffuhGEOOV1z5dnE)


یه سوال چطوری phpmyadmin نصب کردی 
من نصب میکنم موقع لاگین ارور  mysqli_real_connect(): (HY000/2002): No such file or directory میگیرم

**Reply by ItsOrv** [reply]
توی تلگرام راهنماییت میکنم

**Comment by ItsOrv** [comment]
خب خب
بعد از اضافه شدن دستور  "marzban cli admin list" مشکل حل شد، البته هنوز یکم کمتر نشون میده ولی اونقدر تفاوت زیاد نیست که مشکلی ایجاد کنه
اگر هم نیاز به محاسبه دقیقتر دارین به نظرم با این روش https://t.me/marzhelp/9 از طریق صفحه وب ادمین حجم دقیق رو ببینید
